File: clients/vistest.mjs
/* VisTest - self-hosted visual regression testing.
 * Copyright (C) 2026 Kirill Kulagin
 * SPDX-License-Identifier: AGPL-3.0-or-later
 *
 * This file is part of VisTest. See LICENSE for the full terms and NOTICE for
 * the trademark and commercial-licensing terms. Removing this header does not
 * remove those obligations.
 */

/**
 * VisTest — клиент для Node 18+. Без зависимостей (используется встроенный fetch).
 *
 * Подключает к движку VisTest любые уже написанные JS-тесты:
 * Playwright, Puppeteer, WebdriverIO, Cypress (через task), TestCafe.
 *
 *   import { VisTest } from './clients/vistest.mjs';
 *   const vt = new VisTest({ apiUrl: 'http://localhost:8420' });
 *
 *   // Playwright / Puppeteer
 *   await vt.checkPage(page, 'checkout.png');
 *
 *   // Любой другой источник — уже готовый буфер PNG
 *   await vt.check('checkout.png', pngBuffer);
 *
 *   // В конце прогона — одна запись в истории вместо россыпи проверок
 *   await vt.finish();
 *
 * Стабилизация страницы (заморозка анимаций, детерминированные Math.random и
 * Date, ожидание шрифтов и lazy-load, DOM-снепшот) выполняется тем же самым
 * скриптом, что и в Python-раннере: он забирается с `/api/stabilize.js`.
 * Благодаря этому снимки из JS-тестов сопоставимы с питоновскими.
 */

export class VisualMismatchError extends Error {
  constructor(result) {
    super(VisTest.formatResult(result));
    this.name = 'VisualMismatchError';
    this.result = result;
  }
}

export class VisTest {
  /**
   * @param {object} opts
   * @param {string} [opts.apiUrl]   адрес сервиса, по умолчанию VISTEST_API_URL
   * @param {string} [opts.project]
   * @param {string} [opts.platform] ключ платформы; по умолчанию решает сервер
   * @param {string} [opts.browser]
   * @param {string} [opts.runKey]   объединяет проверки в один прогон
   * @param {string} [opts.token]    токен приёма (X-VisTest-Token), по умолчанию VISTEST_TOKEN
   * @param {boolean}[opts.throwOnFail] бросать исключение при регрессе (по умолчанию да)
   * @param {number} [opts.stabilityShots] сколько кадров слать для детекта динамики
   * @param {object} [opts.options]  переопределение порогов
   */
  constructor(opts = {}) {
    this.apiUrl = (opts.apiUrl || process.env.VISTEST_API_URL || 'http://127.0.0.1:8420')
      .replace(/\/$/, '');
    this.project = opts.project || process.env.VISTEST_PROJECT || 'default';
    this.platform = opts.platform || process.env.VISTEST_PLATFORM || undefined;
    this.browser = opts.browser || 'chromium';
    this.runKey = opts.runKey || process.env.VISTEST_RUN_ID || undefined;
    // Токен приёма. Без него инсталляция с пользователями отвечает 401 на
    // каждый снимок: сессионной куки у раннера в CI нет и быть не может.
    this.token = opts.token || process.env.VISTEST_TOKEN
      || process.env.VISTEST_INGEST_TOKEN || '';
    this.throwOnFail = opts.throwOnFail !== false;
    this.stabilityShots = opts.stabilityShots ?? 3;
    this.stabilityDelayMs = opts.stabilityDelayMs ?? 120;
    this.preset = opts.preset;
    this.options = opts.options;
    this.results = [];
    this._stabilizeJs = null;
  }

  /* ------------------------------------------------------------------ */
  /** Заголовки запроса: токен приёма, если он задан. */
  headers() {
    return this.token ? { 'X-VisTest-Token': this.token } : {};
  }

  /**
   * Завершить прогон: собрать проверки в одну запись истории.
   *
   * До этого каждая проверка жила сама по себе. Вердикт возвращался, картинки
   * складывались на диск — но в списке прогонов не появлялось ничего, то есть
   * ни истории по снимку, ни очереди на ревью, ни ответа «этот билд сломал
   * три экрана». Прогон — отдельная сущность, и создать её может только тот,
   * кто знает, что проверки кончились: сам набор.
   *
   * Вызывать один раз в конце (globalTeardown у Playwright, after у Cypress).
   * Без `runKey` звать нечего: прогон нечем назвать.
   */
  async finish(extra = {}) {
    if (!this.runKey) {
      throw new Error('finish() требует runKey: им прогон и назван');
    }
    if (!this.results.length) return null;

    const first = this.results[0] || {};
    const s = this.summary();
    const body = {
      run_id: this.runKey,
      project: this.project,
      platform: this.platform || first.platform,
      browser: this.browser,
      created_at: extra.startedAt || new Date().toISOString(),
      ci_url: extra.ciUrl || process.env.CI_JOB_URL
        || process.env.BUILD_URL || undefined,
      git: extra.git || gitFromEnv(),
      totals: {
        total: s.total, failed: s.failed, new: s.created,
        passed: s.total - s.failed - s.created,
      },
      comparisons: this.results.map(r => ({
        name: r.name, verdict: r.verdict, platform: r.platform,
        browser: r.browser || this.browser, metrics: r.metrics || {},
        regions: r.regions || [], artifacts: r.artifacts || {},
        size: r.size, notes: r.notes || [],
      })),
    };

    const res = await fetch(`${this.apiUrl}/api/runs`, {
      method: 'POST',
      headers: { ...this.headers(), 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!res.ok) throw new Error(`VisTest API ${res.status}: ${await res.text()}`);
    return res.json();
  }

  /** Отправить готовый PNG (Buffer/Uint8Array/Blob). */
  async check(name, image, extra = {}) {
    const form = new FormData();
    form.append('name', name);
    form.append('image', toBlob(image), 'actual.png');
    for (const f of extra.frames || []) form.append('frames', toBlob(f), 'frame.png');
    if (extra.dom) {
      form.append('dom', new Blob([JSON.stringify(extra.dom)],
        { type: 'application/json' }), 'dom.json');
    }
    form.append('project', this.project);
    form.append('browser', extra.browser || this.browser);
    if (this.platform) form.append('platform', this.platform);
    if (this.runKey) form.append('run_key', this.runKey);
    if (extra.updateBaseline) form.append('update_baseline', 'true');
    if (this.preset) form.append('preset', this.preset);
    const opts = { ...(this.options || {}), ...(extra.options || {}) };
    if (Object.keys(opts).length) form.append('options', JSON.stringify(opts));

    const res = await fetch(`${this.apiUrl}/api/check`,
      { method: 'POST', body: form, headers: this.headers() });
    if (!res.ok) {
      throw new Error(`VisTest API ${res.status}: ${await res.text()}`);
    }
    const result = await res.json();
    this.results.push(result);

    if (result.verdict === 'fail' && this.throwOnFail) {
      throw new VisualMismatchError(result);
    }
    return result;
  }

  /**
   * Снять страницу Playwright/Puppeteer и проверить.
   * Делает всё то же, что питоновский раннер: стабилизирует, снимает N кадров,
   * забирает DOM-снепшот и отправляет всё это на сравнение.
   */
  async checkPage(page, name, opts = {}) {
    await this.stabilize(page, opts);

    const shot = () => (opts.selector
      ? page.locator
        ? page.locator(opts.selector).screenshot()
        : page.$(opts.selector).then(el => el.screenshot())
      : page.screenshot({ fullPage: opts.fullPage !== false }));

    const frames = [];
    const total = opts.stabilityShots ?? this.stabilityShots;
    for (let i = 0; i < Math.max(1, total); i++) {
      if (i) await sleep(this.stabilityDelayMs);
      frames.push(await shot());
    }

    let dom = null;
    if (opts.dom !== false && !opts.selector) {
      dom = await evaluate(page, '() => window.__vistest && window.__vistest.dom()');
    }

    return this.check(name, frames[0], {
      frames: frames.slice(1),
      dom: dom || undefined,
      ...opts,
    });
  }

  /** Инжектировать в страницу тот же скрипт стабилизации, что у Python-раннера. */
  async stabilize(page, opts = {}) {
    if (!this._stabilizeJs) {
      const r = await fetch(`${this.apiUrl}/api/stabilize.js`);
      if (!r.ok) throw new Error(`не удалось получить stabilize.js: ${r.status}`);
      this._stabilizeJs = await r.text();
    }
    await evaluate(page, this._stabilizeJs);
    await evaluate(page,
      `() => window.__vistest.settle({ scroll: ${opts.fullPage !== false} })`);
    await sleep(150);
  }

  /**
   * Инжектировать стабилизацию ДО загрузки страницы — так надёжнее:
   * анимации не успевают отыграть первые кадры, Math.random и Date
   * подменяются до того, как их вызовет приложение.
   * Вызывать один раз при создании контекста.
   */
  async installOnContext(context) {
    if (!this._stabilizeJs) {
      const r = await fetch(`${this.apiUrl}/api/stabilize.js`);
      this._stabilizeJs = await r.text();
    }
    if (context.addInitScript) await context.addInitScript(this._stabilizeJs);
    else if (context.evaluateOnNewDocument) await context.evaluateOnNewDocument(this._stabilizeJs);
  }

  /* ------------------------------------------------------------------ */
  static formatResult(r) {
    const m = r.metrics || {};
    const lines = [
      `Visual mismatch: ${r.name}`,
      `  SSIM=${num(m.ssim_global, 5)}  dE00 mean=${num(m.de_mean)} p95=${num(m.de_p95)}`,
      `  changed area=${num(m.changed_area_pct, 3)}%  max severity=${num(m.max_severity, 1)}`,
    ];
    for (const reg of (r.regions || []).slice(0, 10)) {
      const bits = [];
      if (reg.kind === 'moved') bits.push(`shift=(${reg.moved_dx},${reg.moved_dy})`);
      if (reg.selector) bits.push(reg.selector);
      if (reg.element_text) bits.push(`"${reg.element_text}"`);
      if (reg.caption) bits.push(reg.caption);
      lines.push(`    [${reg.kind}] sev=${num(reg.severity, 1)} ` +
        `@(${reg.x},${reg.y}) ${reg.w}x${reg.h}` + (bits.length ? '  ' + bits.join(' | ') : ''));
    }
    for (const n of r.notes || []) lines.push(`  note: ${n}`);
    if (r.artifacts && r.artifacts.boxes) {
      lines.push(`  разметка: ${r.artifacts.boxes}`);
    }
    return lines.join('\n');
  }

  summary() {
    const failed = this.results.filter(r => r.verdict === 'fail');
    return {
      total: this.results.length,
      failed: failed.length,
      created: this.results.filter(r => r.verdict === 'new_baseline').length,
      names: failed.map(r => r.name),
    };
  }
}

/* -------------------------------------------------------------------- */
/** Ветка и коммит из переменных CI. Пусто — тоже честный ответ. */
function gitFromEnv() {
  const e = process.env;
  const branch = e.VISTEST_BRANCH || e.GITHUB_HEAD_REF || e.GITHUB_REF_NAME
    || e.CI_COMMIT_REF_NAME || e.BRANCH_NAME || '';
  const sha = e.VISTEST_COMMIT || e.GITHUB_SHA || e.CI_COMMIT_SHA
    || e.GIT_COMMIT || '';
  return branch || sha ? { branch, sha } : {};
}

function toBlob(data) {
  if (typeof Blob !== 'undefined' && data instanceof Blob) return data;
  return new Blob([data], { type: 'image/png' });
}

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

function num(v, d = 2) {
  return v == null ? '—' : Number(v).toFixed(d);
}

/** Playwright: evaluate(fnString). Puppeteer: то же. WebdriverIO: execute. */
async function evaluate(page, script) {
  if (page.evaluate) return page.evaluate(script);
  if (page.execute) return page.execute(script);
  throw new Error('Не понимаю объект страницы: нет evaluate/execute');
}

export default VisTest;

File: clients/java/VisTest.java
// VisTest - self-hosted visual regression testing.
// Copyright (C) 2026 Kirill Kulagin
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This file is part of VisTest. See LICENSE for the full terms and NOTICE for
// the trademark and commercial-licensing terms. Removing this header does not
// remove those obligations.

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

/**
 * VisTest client for the JVM. One file, no dependencies, Java 11+.
 *
 * <p>Why one file and not a Maven artifact. A Java agent that replaces the
 * comparison inside somebody else's suite is the most expensive thing on the
 * roadmap and the least verifiable: every suite hooks its screenshots
 * differently, and an agent that guesses wrong fails silently. A dependency in
 * their {@code pom.xml} is also not free — it is a review, a version, a
 * security sign-off. A file you can read in five minutes and drop into
 * {@code src/test/java} is the amount of change a team actually accepts.
 *
 * <p>For a JVM suite this is the SECOND choice, and that is deliberate. The
 * first is to let the suite run where it already runs and hand VisTest the
 * folder afterwards — {@code vistest project ingest <key> --dir target/screenshots}.
 * That needs no code in their repository at all. Use this client when the
 * verdict has to be available inside the test, at the moment of the check.
 *
 * <pre>{@code
 * VisTest vt = new VisTest(System.getenv("VISTEST_API_URL"))
 *         .project("web-e2e")
 *         .token(System.getenv("VISTEST_TOKEN"))
 *         .runKey(System.getenv("CI_JOB_ID"));
 *
 * VisTest.Result r = vt.check("checkout.png", screenshotBytes);
 * assertTrue(r.message(), r.passed);
 * ...
 * vt.finish();   // one run in the history, not a scatter of checks
 * }</pre>
 *
 * <p>Also runnable straight from a shell, without compiling anything:
 * <pre>{@code
 * java VisTest.java http://localhost:8420 web-e2e checkout.png shot.png
 * }</pre>
 */
public final class VisTest {

    private final String apiUrl;
    private final HttpClient http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(15))
            .build();

    private String project = "default";
    private String browser = "chromium";
    private String token = "";
    private String runKey = "";
    private final List<String> bodies = new ArrayList<>();
    private int failed = 0;
    private int created = 0;
    private String platform = "";

    public VisTest(String apiUrl) {
        if (apiUrl == null || apiUrl.isBlank()) {
            apiUrl = "http://127.0.0.1:8420";
        }
        this.apiUrl = apiUrl.endsWith("/")
                ? apiUrl.substring(0, apiUrl.length() - 1) : apiUrl;
    }

    public VisTest project(String value) { this.project = nz(value, project); return this; }
    public VisTest browser(String value) { this.browser = nz(value, browser); return this; }
    public VisTest token(String value) { this.token = nz(value, ""); return this; }
    public VisTest runKey(String value) { this.runKey = nz(value, ""); return this; }

    /** The verdict of one snapshot. */
    public static final class Result {
        public final String name;
        public final String verdict;   // pass | fail | new_baseline | error
        public final boolean passed;
        public final String raw;       // the whole response, for the message

        Result(String name, String verdict, boolean passed, String raw) {
            this.name = name;
            this.verdict = verdict;
            this.passed = passed;
            this.raw = raw;
        }

        /** Something a failing assertion can print without hiding the cause. */
        public String message() {
            return "VisTest: " + name + " -> " + verdict
                    + (passed ? "" : "\n" + raw);
        }
    }

    /** Compare a PNG against the baseline VisTest holds for this name. */
    public Result check(String name, byte[] png) throws IOException, InterruptedException {
        Multipart form = new Multipart()
                .field("name", name)
                .field("project", project)
                .field("browser", browser);
        if (!runKey.isEmpty()) {
            form.field("run_key", runKey);
        }
        form.file("image", "actual.png", "image/png", png);

        HttpRequest.Builder request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/api/check"))
                .timeout(Duration.ofSeconds(120))
                .header("Content-Type", form.contentType())
                .POST(HttpRequest.BodyPublishers.ofByteArray(form.body()));
        if (!token.isEmpty()) {
            request.header("X-VisTest-Token", token);
        }

        HttpResponse<String> response =
                http.send(request.build(), HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 300) {
            throw new IOException("VisTest API " + response.statusCode() + ": "
                    + response.body());
        }

        String raw = response.body();
        bodies.add(raw);
        if (platform.isEmpty()) {
            platform = string(raw, "platform");
        }
        String verdict = string(raw, "verdict");
        if ("fail".equals(verdict)) {
            failed++;
        } else if ("new_baseline".equals(verdict)) {
            created++;
        }
        return new Result(string(raw, "name"), verdict, !"fail".equals(verdict), raw);
    }

    public Result check(String name, Path png) throws IOException, InterruptedException {
        return check(name, Files.readAllBytes(png));
    }

    /**
     * Close the run: turn the checks into one entry in the history.
     *
     * <p>Without it every check lives on its own — a verdict comes back,
     * pictures land on disk, and nothing appears in the run list. No history
     * for the snapshot, no review queue, no answer to "this build broke three
     * screens". Only the suite knows when the checks are over, so only the
     * suite can say so. Needs {@code runKey}: that is what names the run.
     */
    public void finish() throws IOException, InterruptedException {
        if (runKey.isEmpty()) {
            throw new IllegalStateException("finish() needs a runKey: it names the run");
        }
        if (bodies.isEmpty()) {
            return;
        }

        int total = bodies.size();
        StringBuilder json = new StringBuilder(256 + total * 512);
        json.append("{\"run_id\":\"").append(escape(runKey))
            .append("\",\"project\":\"").append(escape(project))
            .append("\",\"browser\":\"").append(escape(browser))
            .append("\",\"platform\":\"").append(escape(platform))
            .append("\",\"git\":{\"branch\":\"").append(escape(env("VISTEST_BRANCH",
                    env("CI_COMMIT_REF_NAME", env("GITHUB_REF_NAME", "")))))
            .append("\",\"sha\":\"").append(escape(env("VISTEST_COMMIT",
                    env("CI_COMMIT_SHA", env("GITHUB_SHA", "")))))
            .append("\"},\"totals\":{\"total\":").append(total)
            .append(",\"failed\":").append(failed)
            .append(",\"new\":").append(created)
            .append(",\"passed\":").append(total - failed - created)
            .append("},\"comparisons\":[");
        // Ответы сервиса вставляются как есть. Собирать их заново значило бы
        // писать JSON-сериализатор ради данных, которые уже пришли валидным
        // JSON'ом — и терять на этом регионы, метрики и ссылки на артефакты.
        for (int i = 0; i < total; i++) {
            if (i > 0) {
                json.append(',');
            }
            json.append(bodies.get(i));
        }
        json.append("]}");

        HttpRequest.Builder request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/api/runs"))
                .timeout(Duration.ofSeconds(60))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json.toString(),
                        StandardCharsets.UTF_8));
        if (!token.isEmpty()) {
            request.header("X-VisTest-Token", token);
        }

        HttpResponse<String> response =
                http.send(request.build(), HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 300) {
            throw new IOException("VisTest API " + response.statusCode() + ": "
                    + response.body());
        }
    }

    public int failedCount() { return failed; }
    public int totalCount() { return bodies.size(); }

    // ---------------------------------------------------------------- //
    //  Разбор ответа
    //
    //  Достаются ровно три строковых поля верхнего уровня, и на этом всё.
    //  Тащить JSON-библиотеку в чужой pom.xml ради `verdict` — это ревью,
    //  версия и согласование безопасности вместо одной строки; полный ответ
    //  всё равно уезжает в сообщение об ошибке, где он и нужен целиком.
    // ---------------------------------------------------------------- //
    static String string(String json, String key) {
        String needle = "\"" + key + "\"";
        int at = json.indexOf(needle);
        if (at < 0) {
            return "";
        }
        // Пробелы между ключом, двоеточием и значением допустимы по формату, и
        // тот, кто отдаёт ответ, волен их ставить: прокси, другой сериализатор,
        // «красивый» JSON в логе. Привязка к `"key":"` работала ровно с одним
        // из этих вариантов и молча возвращала пустую строку на остальных —
        // то есть вердикт становился пустым, а снимок «прошедшим».
        int i = at + needle.length();
        while (i < json.length() && Character.isWhitespace(json.charAt(i))) {
            i++;
        }
        if (i >= json.length() || json.charAt(i) != ':') {
            return "";
        }
        i++;
        while (i < json.length() && Character.isWhitespace(json.charAt(i))) {
            i++;
        }
        if (i >= json.length() || json.charAt(i) != '"') {
            return "";
        }
        StringBuilder out = new StringBuilder();
        for (int at2 = i + 1; at2 < json.length(); at2++) {
            char c = json.charAt(at2);
            if (c == '\\' && at2 + 1 < json.length()) {
                out.append(json.charAt(++at2));
                continue;
            }
            if (c == '"') {
                break;
            }
            out.append(c);
        }
        return out.toString();
    }

    private static String escape(String value) {
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static String nz(String value, String fallback) {
        return value == null || value.isBlank() ? fallback : value;
    }

    private static String env(String name, String fallback) {
        String value = System.getenv(name);
        return value == null || value.isBlank() ? fallback : value;
    }

    // ---------------------------------------------------------------- //
    //  multipart/form-data вручную: в стандартной библиотеке его нет
    // ---------------------------------------------------------------- //
    static final class Multipart {
        private final String boundary = "vistest-" + Long.toHexString(System.nanoTime());
        private final ByteArrayOutputStream out = new ByteArrayOutputStream();

        Multipart field(String name, String value) {
            write("--" + boundary + "\r\nContent-Disposition: form-data; name=\""
                    + name + "\"\r\n\r\n" + value + "\r\n");
            return this;
        }

        Multipart file(String name, String filename, String type, byte[] data) {
            write("--" + boundary + "\r\nContent-Disposition: form-data; name=\""
                    + name + "\"; filename=\"" + filename + "\"\r\nContent-Type: "
                    + type + "\r\n\r\n");
            out.write(data, 0, data.length);
            write("\r\n");
            return this;
        }

        String contentType() {
            return "multipart/form-data; boundary=" + boundary;
        }

        byte[] body() {
            write("--" + boundary + "--\r\n");
            return out.toByteArray();
        }

        private void write(String text) {
            byte[] bytes = text.getBytes(StandardCharsets.UTF_8);
            out.write(bytes, 0, bytes.length);
        }
    }

    // ---------------------------------------------------------------- //
    public static void main(String[] args) throws Exception {
        if (args.length < 4) {
            System.err.println(
                    "usage: java VisTest.java <api-url> <project> <name> <file.png>\n"
                  + "       token is taken from VISTEST_TOKEN, run key from VISTEST_RUN_ID");
            System.exit(2);
        }
        VisTest vt = new VisTest(args[0])
                .project(args[1])
                .token(System.getenv("VISTEST_TOKEN"))
                .runKey(System.getenv("VISTEST_RUN_ID"));
        Result r = vt.check(args[2], Path.of(args[3]));
        System.out.println(r.message());
        if (!vt.runKey.isEmpty()) {
            vt.finish();
        }
        System.exit(r.passed ? 0 : 1);
    }
}
